apps/labs/posts/python-wheels-from-tags-to-variants.md

```diff
@@ -0,0 +1,1146 @@
+
+There are two dependency-related problems that could be relevant to variant
+support: expressing dependencies that are specific to a subset of variants,
+and requiring a specific variant (or a subset of variants) of another package.
+So far, we have deferred working on the latter problem, as we did not have a specific
+use case to focus the design on, and we did not want to arrive at a complex
+solution that would not necessarily match what the users actually needed.
+On the other hand, variant-specific dependencies had immediate use cases.
+
+The idea of variant-specific dependencies arrived into the project quite early.
+After all, building a variant wheel generally involves installing a provider
+plugin, if only to validate whether the requested properties are correct.
+And indeed, the very first attempt to do that used environment markers
+in the `build-system.requires` section of `pyproject.toml`, to specify which
+plugins needed to be installed. It looked somewhat like the following:
+
+<figure>
+
+```toml
+[build-system]
+build-backend = "mesonpy"
+requires = [
+  "meson-python",
+  "provider-variant-x86-64; 'x86_64' in variant_namespaces",
+  "provider-variant-aarch64; 'aarch64' in variant_namespaces",
+]
+```
+
+<figcaption>Listing 8. Example `pyproject.toml` with variant provider plugins selected via environment markers</figcaption>
+</figure>
+
+These markers meant that when the variant wheel was built with a property
+in the listed namespace, the relevant provider plugin was installed
+in the build environment, and therefore became available to the build backend.
+However, this solution did not last long. It was quite problematic to implement
+given the limited interface between [pypa/build](https://pypi.org/project/build/)
+and the installers used by it.
+Besides, provider plugins also needed to be installed while installing variant wheels, so it made
+more sense to separate them, as described earlier in the post.
+
+Nevertheless, the syntax based on environment markers made sense, and it eventually made
+its way into the actual package dependencies. A package that previously had
+to dynamically declare different dependency sets for every wheel variant,
+now could instead use them to apply dependencies conditionally to the selected
+variant.
+
+<figure>
+
+```toml
+dependencies = [
+    'triton==3.4.0; platform_system == "Linux" and "nvidia" in variant_namespaces',
+    'nvidia-cudnn==9.10.2.21; platform_system == "Linux" and platform_machine == "x86_64" and "nvidia" in variant_namespaces',
+    'nvidia-cusparselt==0.7.1; platform_system == "Linux" and platform_machine == "x86_64" and "nvidia" in variant_namespaces',
+    'nvidia-nccl==2.27.3; platform_system == "Linux" and platform_machine == "x86_64" and "nvidia" in variant_namespaces',
+
+    # CUDA 12.6
+    'nvidia-cuda-nvrtc==12.6.77; platform_system == "Linux" and platform_machine == "x86_64" and "nvidia :: cuda_version_lower_bound :: 12.0" in variant_properties',
```

Comment:
Is the divergence between 12.6 and 12.0 here intentional?

Reply:
Yep, 12.6 builds are used for `>=12.0,<12.8`. Perhaps I should make it clear above.